indexed_filesource.go

```diff
@@ -0,0 +1,197 @@
+		nextSourceFactory:  nextSourceFactory,
+		nextHandlerWrapper: nextHandlerWrapper,
+	}
+
+}
+
+type IndexedFileSource struct {
+	*shutter.Shutter
+
+	logger        *zap.Logger
+	handler       Handler
+	lastProcessed BlockRef
+
+	blockIndex BlockIndex
+	blockStore dstore.Store
+
+	nextSourceFactory  SourceFromNumFactory
+	nextHandlerWrapper func(h Handler, lib BlockRef) Handler
+
+	preprocFunc PreprocessFunc
+}
+
+func (s *IndexedFileSource) Run() {
+	s.Shutdown(s.run())
+}
+
+var SkipToNextRange = errors.New("skip to next range")
+var NoMoreIndex = errors.New("no more index")
+
+type Skippable struct{}
+
+var SkipThisBlock = Skippable{}
+
+func (s *IndexedFileSource) SetLogger(l *zap.Logger) {
+	s.logger = l
+}
+
+func (s *IndexedFileSource) run() error {
+	for {
+		base, lib, hasIndex := s.blockIndex.NextBaseBlock()
+		if !hasIndex {
+			libString := ""
+			if lib != nil {
+				libString = lib.String()
+			}
+			s.logger.Debug("indexed file source switching to next source", zap.Uint64("base", base), zap.String("lib", libString))
+			nextHandler := s.nextHandlerWrapper(s.handler, lib)
+			nextSource := s.nextSourceFactory(base, nextHandler)
+			nextSource.OnTerminated(func(err error) {
+				s.Shutdown(err)
+			})
+			nextSource.Run()
+			return nextSource.Err()
+		}
+
+		s.logger.Debug("indexed file source starting a file source, backed by index", zap.Uint64("base", base))
+
+		fs := NewFileSource(s.blockStore, base, 1, s.wrappedPreproc(), s.wrappedHandler())
+		s.OnTerminating(func(err error) {
+			fs.Shutdown(err)
+		})
+
+		err := fs.run()
+
+		if errors.Is(err, SkipToNextRange) {
+			continue
+		}
+		if errors.Is(err, NoMoreIndex) {
+			continue
+		}
+		return err
+	}
+}
+
+func (s *IndexedFileSource) wrappedPreproc() PreprocessFunc {
+	return func(blk *Block) (interface{}, error) {
+		if s.blockIndex.Skip(blk) {
+			return SkipThisBlock, nil
+		}
+		if s.preprocFunc == nil {
+			return nil, nil
+		}
+		return s.preprocFunc(blk)
+	}
+}
+
+func safeMinus(i, j uint64) uint64 {
+	if i < j {
+		return 0
+	}
+	return i - j
+}
+
+func (s *IndexedFileSource) wrappedHandler() Handler {
+	var skipCount uint64
+	return HandlerFunc(func(blk *Block, obj interface{}) error {
+		if _, ok := obj.(Skippable); ok { // SkipThisBlock from wrappedPreproc
+			skipCount++
+			if skipCount%10 == 0 {
+				nextBase, _, hasIndex := s.blockIndex.NextBaseBlock()
+				if hasIndex && safeMinus(blk.Number, nextBase) > 200 {
+					return SkipToNextRange
+				}
+			}
+			return nil
+		}
+
+		ppblk := &PreprocessedBlock{
+			blk,
+			obj,
+		}
+		toProc, indexedRangeComplete := s.blockIndex.Reorder(ppblk)
+
+		for _, ppblk := range toProc { // sending New, Irreversible for each block
+			if err := s.handler.ProcessBlock(ppblk.Block, wrapIrreversibleBlockWithCursor(blk, ppblk.Obj, StepNew)); err != nil {
+				return err
+			}
+			if err := s.handler.ProcessBlock(ppblk.Block, wrapIrreversibleBlockWithCursor(blk, ppblk.Obj, StepIrreversible)); err != nil {
```

Comment:
Since this is a hot path, maybe dedupe the creation of that `wrapIrreversibleBlockWithCursor`  and go simply modify the `step` when calling the next time.

Reply:
There's a few allocations, and some `AsRef()` that we really don't need to do twice, nor even three times.

Reply:
I'm worried about "modifying" the step on the object, (if something buffers it later...) I'll reduce the calls to AsRef() from 6 to 1